src/storage/txn/txn_status_cache.rs

```diff
@@ -301,93 +380,185 @@ impl TxnStatusCache {
+                    after_allocated = large_txn_cache.internal_allocated_capacity();
+                } else {
+                    let mut normal_cache = self.normal_cache[slot_index].lock();
+                    previous_size = normal_cache.size();
+                    previous_allocated = normal_cache.internal_allocated_capacity();
+                    normal_cache.insert(start_ts, entry);
+                    after_size = normal_cache.size();
+                    after_allocated = normal_cache.internal_allocated_capacity();
+                }
+            }
+            TxnState::Committed(_) | TxnState::RolledBack => {
+                let mut normal_cache = self.normal_cache[slot_index].lock();
+                previous_size = normal_cache.size();
+                previous_allocated = normal_cache.internal_allocated_capacity();
+                normal_cache.insert(start_ts, entry);
+                after_size = normal_cache.size();
+                after_allocated = normal_cache.internal_allocated_capacity();
+
+                let mut large_cache = self.large_txn_cache[slot_index].lock();
+                if let Some(existing_entry) = large_cache.get_mut(&start_ts) {
+                    // don't update committed or rolled back txns.
+                    if let TxnState::Ongoing(_) = existing_entry.state {
+                        existing_entry.state = state;
+                        existing_entry.update_time = update_time;
+                    }
+                }
+            }
+        }
         SCHED_TXN_STATUS_CACHE_SIZE
             .used
-            .add(size as i64 - previous_size as i64);
+            .add(after_size as i64 - previous_size as i64);
         SCHED_TXN_STATUS_CACHE_SIZE
             .allocated
-            .add(allocated as i64 - previous_allocated as i64);
+            .add(after_allocated as i64 - previous_allocated as i64);
+    }
+
+    /// Insert a committed txn into the normal cache.
+    pub fn insert_committed(&self, start_ts: TimeStamp, commit_ts: TimeStamp, now: SystemTime) {
+        let state = TxnState::Committed(commit_ts);
+        self.upsert(start_ts, state, now);
+    }
+
+    pub fn get(&self, start_ts: TimeStamp) -> Option<TxnState> {
+        if !self.is_enabled {
+            return None;
+        }
+
+        let slot_index = self.slot_index(start_ts);
+
+        // large txn cache should be queried first, because it has a higher priority
+        // when upserting. It's possible for a start_ts to exist in both caches.
+        // The one in large_txn_cache is more up-to-date.
+        let mut large_txn_cache = self.large_txn_cache[slot_index].lock();
+        if let Some(entry) = large_txn_cache.get(&start_ts) {
+            return Some(entry.state.clone());
+        }
+
+        let mut normal_cache = self.normal_cache[slot_index].lock();
+        if let Some(entry) = normal_cache.get(&start_ts) {
+            return Some(entry.state.clone());
+        }
```

Comment:
It's definitely possible. But I think it would be less robust against worst case scenario.
Btw the two caches can be viewed as two individual parts and they do not depend on each other. So I suppose there should be no risks here?